### ISSUE TITLE
Strange behavior with display driver

### ISSUE THREAD

**Issue by MikeAlhayek** [OrchardCMS/OrchardCore#14250]
The new [UserMenuDisplayDriver](https://github.com/OrchardCMS/OrchardCore/blob/b16d29b002eeb547cf26be280b76832b4b093e4d/src/OrchardCore.Modules/OrchardCore.Users/Drivers/UserMenuDisplayDriver.cs#L28) creates a new shape `UserMenuItems__Title` that gets injected into the `Header` location on both `Detail` and `DetailAdmin` display modes. 

This work with no issue on the source code. However, I am using preview packages on my project and trying to change that shape.

In my project I want to change the default title. So I added this driver

```
public class DisplayNameUserMenuDisplayDriver : DisplayDriver<UserMenu>
{
    public override IDisplayResult Display(UserMenu model)
    {
        var t = View("UserMenuItems__DisplayName", model)
            .Location("Detail", "Header:4")
            .Location("DetailAdmin", "Header:4");

        return Combine(t);
    }
}
````

When I run my project, I indeed see the content of the `UserMenuItems-DisplayName.cshtml` on the menu header as expected. 

As you can see here, **NEW VALUE!** is injected on the header section as expected.
![image](https://github.com/OrchardCMS/OrchardCore/assets/24724371/dd733d6c-cbcf-45d6-8cf3-86f4cfe1d7ed)

Now, I am trying to stop rendering the `UserMenuItems__Title` since my `UserMenuItems__DisplayName` has a replacement shape. So I added a `placement.json` file with the following content

```json
{
  "UserMenuItems__Title": [
    {
      "place": "-"
    }
  ]
}
```
But for some odd reason, the the entries shape no longer renders. I am expecting that the shape `UserMenuItems__Title` will not be rendered by everything else including my new shape will. 

You see how the menu is broken "it does not even display
![image](https://github.com/OrchardCMS/OrchardCore/assets/24724371/80a9cffa-3186-4ff7-96cc-6e7a6d7e2875)

Any idea what could be causing placement to stop hiding the entire menu instead of the header only?

**Comment by jtkech** [comment]
In the context of the placement the shape type is the first term before any `_, __` delemiter, so here the shape type is `UserMenuItems`, that's why all are hidden.

In your driver you need to add differentiators, for example

            View("UserMenuItems__Title", model)
            .Location("Detail", "Header:5")
            .Location("DetailAdmin", "Header:5")
            .Differentiator("Title"), // <= This one.

Then for example the following will work

    "UserMenuItems": [
      {
        "place": "-",
        "DisplayType": "DetailAdmin",
        "Differentiator": "Title"
      }
    ]

**Comment by MikeAlhayek** [comment]
@jtkech thanks for the info. I think we have to change the drivers for the UserMenu. 

I never saw Differentiator("") is used like that in drivers. 

Not sure what is the best course of action to correct this. Do we consider a single shape "UserMenuItem" and specify Differentiator("Title") or is there a better way? If we use Differentiator("") what will that do to the view names, specially that we are using multiple displayTypes.

**Comment by jtkech** [comment]
That's okay to have these different shape types having the same `UserMenuItems`base shape type.

The differentiator is only used for placements, as we use for content types, parts and fields.

`Differentiator()` is used for us in `ContentFieldDisplayDriver`, `ContentPartDisplayDriver` and `ContentItemDisplayCoordinator` to take into account content types, parts and fields names.

Just saw that it is also used in `CustomUserSettingsDisplayDriver`.

But yes, it may be missing for other generic `DisplayDriver<>` when they drive multiple shape with the same base shape type that would need to be differentiated.

But nobody complained.

**Comment by MikeAlhayek** [comment]
Do we also have to change the shape type from UserMenuItems__{SubType}? Or, can we keep that as is? If we can keep that as is, can't we just identify the differentiator in the ShapeTable by setting it to something like UserMenuItem-{subType}?

**Comment by jtkech** [comment]
Yes I think we can keep that as is.

> can't we just identify the differentiator in the ShapeTable ...?

Not sure I understand

As I can see, in your driver you only need to add the following for each driven shape.

    ...
    .Differentiator("Title"),
    ...
    ...
    ...
    .Differentiator("SignedUser"),
    ...
    ...
    ...
    .Differentiator("Profile"),
    ...
    ...
 And so on

**Comment by MikeAlhayek** [comment]
I have to try it tomorrow. But differentiator is the shape type `UserMenuItems__Title` So I thought one of the following would work but did not

```
{
  "UserMenuItems": [
    {
      "place": "-",
      "differentiator": "UserMenuItems__Title"
    }
  ]
}
```

or 

```
{
  "UserMenuItems__Title": [
    {
      "place": "-",
      "differentiator": "UserMenuItems__Title"
    }
  ]
}
```

> can't we just identify the differentiator in the ShapeTable ...?

I mean changing the `UserMenuShapeTableProvider` to something like this 

```c#

public class CustomUserMenuShapeTableProvider : IShapeTableProvider
{
    private const string ShapePrefix = "UserMenuItems";

    private const string ShapeTypePrefix = $"{ShapePrefix}__";

    private const string ShapeAlternatePrefix = $"{ShapePrefix}_";

    public void Discover(ShapeTableBuilder builder)
    {
        // Describe any shape-type that starts with 'UserMenuItems__'.
        builder.Describe($"{ShapeTypePrefix}*")
            .OnDisplaying(context =>
            {
                // The value of 'subType' is the encoded string that comes after 'UserMenuItems__'.
                var subType = context.Shape.Metadata.Type[ShapeTypePrefix.Length..].EncodeAlternateElement();

                context.Shape.Metadata.Differentiator = subType;

                if (String.IsNullOrEmpty(context.Shape.Metadata.DisplayType) || context.Shape.Metadata.DisplayType == "Detail")
                {
                    return;
                }

                // 'UserMenuItems_{displayType}' e.g. 'UserMenuItems.DetailAdmin.cshtml'.
                context.Shape.Metadata.Alternates.Add($"{ShapeAlternatePrefix}{context.Shape.Metadata.DisplayType}");

                // 'UserMenuItems_{displaType}__{subType}' e.g. 'UserMenuItems-Dashboard.DetailAdmin.cshtml'.
                context.Shape.Metadata.Alternates.Add($"{ShapeAlternatePrefix}{context.Shape.Metadata.DisplayType}__{subType}");
            });
    }
}
```

but even with setting the `differentiator` in the shape table, did not work.

**Comment by jtkech** [comment]
> I have to try it tomorrow. But differentiator is the shape type UserMenuItems__Title

Yes it defaults to the shape type if you don't use `Differentiator()`.


> So I thought one of the following would work but did not

Yes but only if you use `.Differentiator("UserMenuItems__Title")` in your driver, then here better to use `UserMenuItems-Title` with a dash, but not so good to include the base shape type in the differentiator.

I tried to add `Differentiator("Title")` with the placement I mentioned, it works on my side.

> I mean changing the UserMenuShapeTableProvider to something like this

Okay but it's too late.

As I can see all is good, only the `Differentiator("...")` need to be added.
